### ISSUE TITLE
Context support

### ISSUE THREAD

**Issue by arnaud-lb** [cenkalti/backoff#32]
Hi

I would like to request the possibility to support contexts natively in this package, now that Go ships with a context package.

Contexts allow to _cancel_ code as soon as we are not interested in the result anymore. For instance, if a client has a 120 seconds timeout, there is no reason to continue processing a request after 120 seconds.

The Retry and RetryNotify functions, however, will prevent cancellation for the duration specified in MaxElapsedTime. Even if the retried operation cancels early because it notices that some context is done, Retry/RetryNotify will sleep and call the operation again until MaxElapsedTime have actually elapsed.

Currently, the following solutions can be used:

* Setting MaxElapsedTime to a duration that matches the context's expected deadline. However, the deadline of a context isn't known (and not all contexts are deadline-based), so this solution is not practical.
* Using a mix of ticker and select:

    ```
    ticker := context.NewTicker(b)
    defer ticker.Cancel()

    outer_loop:
    for {
        select {
            case <-ticker.C:
                if err := op(); err == nil {
                    break outer_loop;
                }
            case <-ctx.Done():
                break outer_loop;
        }
    }
    ```
    However, this is cumbersome compared to using backoff.Retry()

It would be awesome if this package could support contexts natively.

One possibility would be to add a Context property on the backoff struct. This would be the only visible change to the interface, and would be fully backwards-compatible:

```
b := backoff.NewExponentialBackoff()
b.Context = ctx

context.Retry(op, b) // aborts as soon as context is done, or MaxElapsedTime is elapsed
```

This is fully backwards compatible, and doesn't complexifies the interface or the implementation.

**Comment by cenkalti** [comment]
I like this. I will be happy to merge if you make a PR.

**Comment by arnaud-lb** [comment]
Awesome! Here is a PR: #34

**Comment by gsempe** [comment]
It would be a good improvement of the lib

**Comment by cenkalti** [comment]
Merged #34

**Comment by arnaud-lb** [comment]
Thanks !
